src/series/algorithm/waterfall.js

```diff
@@ -0,0 +1,107 @@
+import d3 from 'd3';
+import {noop} from '../../util/fn';
+
+export default function() {
+
+    var xValueKey = '',
+        yValue = function(d) { return d.y; },
+        startsWithTotal = false,
+        totals = noop;
+
+    var waterfall = function(data) {
+        var length = data.length,
+            i = 0,
+            previousEnd = yValue(data[0]),
+            start,
+            end,
+            total,
+            result = [],
+            finalIndex = length - 1,
+            customTotals = false;
+
+        if (startsWithTotal) {
+            // First value is a total
+            result.push({
+                domain: data[0][xValueKey],
```

Comment:
I'm not keen on the property names that the algorithm outputs, how about

 - domain => x
 - start => y0
 - end => y1

This is similar to the [d3 stack layout](https://github.com/mbostock/d3/wiki/Stack-Layout) datum properties.